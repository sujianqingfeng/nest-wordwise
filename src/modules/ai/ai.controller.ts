import { CurrentUser } from "@/decorator/user.decorator";
import { Controller, Get, Query } from "@nestjs/common";
import { ProfileService } from "@/modules/profile/profile.service";
import { AIService } from "./ai.service";
import { ANALYSIS_GAMMAR_PROMPT } from "@/constants";
import { AnalyticsGrammarDto } from "./dtos/ai.dto";

@Controller("ai")
export class AIController {
  constructor(
    private profileService: ProfileService,
    private aiService: AIService,
  ) {}

  @Get("/analysis-grammar")
  async analysisGrammar(
    @CurrentUser("id") userId: string,
    @Query() query: AnalyticsGrammarDto,
  ) {
    const profile = await this.profileService.profile(userId);

    const { text } = query;
    const prompt = `${ANALYSIS_GAMMAR_PROMPT}${text}`;

    return this.aiService.generateContent(prompt, profile);
  }
}
