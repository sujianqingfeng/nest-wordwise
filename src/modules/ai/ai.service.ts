import type { Profile } from '@/modules/drizzle/types'
import { Injectable } from '@nestjs/common'
import { GeminiProvider } from './providers/genimi.provider'
import { OpenAIProvider } from './providers/openai.provider'

@Injectable()
export class AIService {
  constructor(
    private geminiProvider: GeminiProvider,
    private openAIProvider: OpenAIProvider
  ) {}

  async generateContent(prompt: string, profile: Profile) {
    const { defaultAIEngine, geminiKey, openAIKey } = profile

    let result = null

    switch (defaultAIEngine) {
      case 'gemini':
        result = await this.geminiProvider.generateContent(prompt, geminiKey)
        break
      case 'openAI':
        result = await this.openAIProvider.generateContent(prompt, openAIKey)
        break

      default:
        throw new Error('Unsupported AI engine')
    }

    return {
      result
    }
  }
}
