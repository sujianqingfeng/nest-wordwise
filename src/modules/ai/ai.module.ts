import { Module } from "@nestjs/common";
import { AIService } from "./ai.service";
import { AIController } from "./ai.controller";

@Module({
  providers: [AIService],
  exports: [AIService],
  controllers: [AIController],
})
export class AiModule {}
