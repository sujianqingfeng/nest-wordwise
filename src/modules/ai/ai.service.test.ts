import { HttpModule } from '@nestjs/axios'
import { Test, TestingModule } from '@nestjs/testing'
import { AIService } from './ai.service'
import { GeminiProvider } from './providers/genimi.provider'

describe('AIService', () => {
  let service: AIService

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [HttpModule],
      providers: [AIService, GeminiProvider]
    }).compile()

    service = module.get<AIService>(AIService)
  })

  it('should be defined', () => {
    expect(service).toBeDefined()
  })

  it('connected', async () => {
    const result = await service.generateContent('hello', 'xxx')
    expect(result).toMatchInlineSnapshot(
      `"Hello there. How can I help you today?"`
    )
  })
})
