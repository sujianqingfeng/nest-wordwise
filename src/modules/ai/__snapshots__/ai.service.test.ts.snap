// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AIService open-ai 1`] = `"Hello! How can I assist you today?"`;
